### PR TITLE
feat: restructure op builder

### LINKED ISSUE #11040: Improve OP payload builder
### Describe the feature

this https://github.com/paradigmxyz/reth/pull/10904 attempted to restructure the OP payload builder. this now has a few sub functions that are hard to follow and some redundant arguments.

imo this should be reverted and go through review again.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
supersedes #11073

closes https://github.com/paradigmxyz/reth/issues/11040

This refactors how the OP payload is being constructed.
Introduces a new `Ctx` type that provides reusable logic for transacting.
This now also makes it possible easily modify how new txs are getting included int the block

### CODE REVIEW

**Review by klkvr — APPROVED** [review]
lgtm!

File: crates/optimism/payload/src/builder.rs [hunk 1]
```diff
@@ -152,7 +811,7 @@ where
 /// and configuration, this function creates a transaction payload. Returns
 /// a result indicating success with the payload or an error in case of failure.
 #[inline]
-pub(crate) fn optimism_payload<EvmConfig, Pool, Client>(
+pub fn optimism_payload<EvmConfig, Pool, Client>(
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/optimism/payload/src/builder.rs:819]: yep, will remove before merging

File: crates/optimism/payload/src/builder.rs [hunk 2]
```diff
@@ -134,12 +185,620 @@ where
...
+            let env = EnvWithHandlerCfg::new_with_cfg_env(
+                self.initialized_cfg.clone(),
+                self.initialized_block_env.clone(),
+                self.evm_config.tx_env(tx.as_signed(), tx.signer()),
+            );
+
+            // Configure the environment for the block.
+            let mut evm = self.evm_config.evm_with_env(&mut *db, env);
+
+            let ResultAndState { result, state } = match evm.transact() {
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/optimism/payload/src/builder.rs:750]: yep, I'd like that, but separately, wanted to keep it as similar to the previous version as possible
